Sprint 12

superaydumov opened the pull request:
Pull request according to Yandex Practikum course (sprint 12).

superaydumov committed:
Added ImageListService class to maintain pictures download (also created Photo and PhotoResult models).
Moved DateFormatter code to DateExtension.
Added tableView method to start photos downloading while scrolling.

superaydumov committed:
Little tweaks and mistakes correction.

superaydumov committed:
Refactored fetchPhotoNextPage method and PhotoResult model after working with test in different branch.

superaydumov committed:
Modified ImagesListViewController and ImagesListCell to show downloaded photos in table view:
- updated tableView methods in ImagesListViewController class;
- added updateImagesList and updateTableViewAnimated methods to update content in table view;
- updated configureCell method to work with downloaded photos.

superaydumov committed:
Added changeLike method to ImagesListViewController class.
Added LikePhotoResult structure in PhotoResult model.
Created IBAction likeButtonDidTap in ImagesListCell class.

superaydumov committed:
Refactored ImagesListCell and ImagesListViewController to conform delegate pattern to add likes on photos ( added imagesListCellDidTapLike method).
Modified SingleImageViewController to show full size photo which was downloaded with ImagesListService.

superaydumov committed:
Refactored ProfileViewController to add logout logic (switchToSplashViewController method) and cleaning cookies (cleanWebCookies method).

superaydumov committed:
Added alerts to ImagesListViewController, ProfileViewController and SingleImageViewController.
Added clean methods to ImagesListService, ProfileService and ProfileImageService to erase all user info after logout from the app.
Also little tweaks in SingleImageViewController.

DanilaG commented on ImageFeed/Helpers/String+Extension.swift:
```diff
@@ -0,0 +1,16 @@
+extension String {
+    var dateTimeString: Date? {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ"
```
Давай заменим на `ISO8601DateFormatter`)

DanilaG commented on ImageFeed/Services/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,152 @@
+                    let likedByUser = body.photo?.likedByUser ?? false
+                    if let index = self.photos.firstIndex(where: { $0.id == body.photo?.id}) {
+                        let photo = self.photos[index]
+                        let newphoto = Photo(id: photo.id,
```
У тебя классный код стайл) 
Но название `newphoto` выбивается из него. Давай поправим нзвание на `newPhoto`, чтобы стиль названия не отличался от других переменных

DanilaG commented on ImageFeed/Helpers/Date+Extension.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+var dateFormatter: DateFormatter = {
```
Давай сделаем `let`, чтоы избежать случайного изменения форматера кем-либо

Было бы супер, если создать name space для всех необходимых форматеров, что позволило бы упростить читаймость и работу с автодополнениями. Так, например, можно сделать что-то вроде:
```
/// Кастомный форматеры даты
enum DateFormatters {
    /// Форматер в стиле `4 сентября 2023 г.`
    let long: DateFormatter  = { ... }()
    /// Форматер в стиле `2023-09-08T16:05:44Z`
    let iso8601: DateFormatter = { ... }()
}
```  
И использовать как:
```
DateFormatters.long.string(from: ...)
```

DanilaG commented on ImageFeed/Services/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,152 @@
+        task.resume()
+    }
+    
+    func changelike(photoId: String, isLike: Bool, _ completion: @escaping (Result<Bool, Error>) -> Void) {
```
Давай поправим название с `changelike` на `changeLike`, чтобы оно по стилю совподало с остальным кодом

DanilaG commented on ImageFeed/Services/ImagesList/ImagesListCell.swift:
```diff
@@ -18,18 +18,55 @@ final class ImagesListCell: UITableViewCell {
     // MARK: - Constants
     
-    static let reuseIdentifier = "ImagesListCell"
+    struct Keys {
+        static let reuseIdentifier = "ImagesListCell"
+        static let placeholderImage = "image_placeholder"
+        static let likedButtonOn = "like_button_ON"
+        static let likedButtonOff = "like_button_OFF"
+    }
```
Супер, что названия сохронены в константы. Это позволит не допустить опечаток

DanilaG commented on ImageFeed/Services/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,152 @@
+        assert(Thread.isMainThread)
+        task?.cancel()
+        
+        guard let token = token else { return }
```
Давай добавим вызов `completion`. Так как если данная ошибка появится, то код обрабатывающий результаты - не вызоваится. Что, в теории, может привести к зависающим лоудерам

DanilaG commented on ImageFeed/Services/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,152 @@
+        
+        request?.addValue("Bearer \(token))", forHTTPHeaderField: "Authorization")
+
+        guard let request else { return }
```
Давай добавим вызов `completion`. Так как если данная ошибка появится, то код обрабатывающий результаты - не вызоваится. Что, в теории, может привести к зависающим лоудерам

DanilaG commented on ImageFeed/Services/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,152 @@
+
+        let task = urlSession.objectTask(for: request) { [weak self] (result: Result<LikePhotoResult, Error>) in
+            DispatchQueue.main.async {
+                guard let self else { return }
```
Давай добавим вызов `completion`. Так как если данная ошибка появится, то код обрабатывающий результаты - не вызоваится. Что, в теории, может привести к зависающим лоудерам

DanilaG commented on ImageFeed/ViewControllers/ProfileViewController.swift:
```diff
@@ -80,6 +85,55 @@ final class ProfileViewController: UIViewController {
+    }
+    
+    private func cleanServicesData() {
+        imagesListService.clean()
+        profileService.clean()
+        profileImageService.clean()
```
Клёво, что чистятся не нужные даные, так как это позволяет экономить ресурсы

superaydumov committed:
Refactored code according to reviewer comments:
- created DateFormatters structure with ISO8601DateFormatter and DateFormatter;
- implemented new date formatters in ImagesListCell and ImagesListService to make this operations "cheaper";
- little tweaks and changes according to review.
